sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,437 @@
+    }
+    if (dataAlignSize2 > 0 && dataAlignSize1 > 0) { // both types are variable length
+      return dataAlignSize1 < dataAlignSize2;
+    }
+    return refSize1 > refSize2;
+  }
+
+  /**
+   * Sets the alignment for variable length data types.
+   * This is used to align the data in the variable width section of the JCUDF row.
+   * For fixed-length types, the alignment is equivalent to the type size in bytes.
+   *
+   * @param rapidsType a variable length data type.
+   * @return the alignment in bytes of the RAPIDS type in the variable width JCUDF.
+   */
+  public static int getDataAlignmentForDataType(DType rapidsType) {
+    switch (rapidsType.getTypeId()) {
+      case STRING:
+        // Strings are 1-byte aligned
+        return JCUDF_STRING_TYPE_ALIGNMENT;
+      case LIST:
+        // Data aligned depends on the datatype. return 8 bytes for now.
+      case STRUCT:
+        // Structs are 8-bytes aligned
+        return JCUDF_VAR_LENGTH_DATA_DEFAULT_ALIGNMENT;
+      default:
+        // all other types are aligned based on the size.
+        return rapidsType.getSizeInBytes();
+    }
+  }
+
+  /**
+   * Sets an estimate size in bytes for variable-length data types.
+   * This is used to get a rough estimate of the bytes needed to allocate JCUDF row.
+   *
+   * @param rapidsType a variable length data type.
+   * @return size in bytes of the variable length datatype.
+   */
+  public static int getEstimateSizeForVarLengthTypes(DType rapidsType) {
+    switch (rapidsType.getTypeId()) {
+      case STRING:
+        return JCUDF_TYPE_STRING_LENGTH_ESTIMATE;
+      case LIST:
+      case STRUCT:
+        // Return 32 bytes for other types until we decide on how to handle each type.
+      default:
+        throw new IllegalArgumentException(rapidsType + " estimated size is not supported yet.");
+    }
+  }
+
+  /**
+   * Checks if CUDF Kernel can apply optimized row conversion on the Spark schema represented by
+   * {@link CudfUnsafeRow}. Note that the calculation assumes that the {@code cudfUnsafeRow}
+   * is not usable yet (i.e., {@link CudfUnsafeRow#pointTo(long, int)} is not called yet).
+   *
+   * The fixed-width optimized cudf kernel only supports up to 1.5 KB per row.
+   * For fixed-size rows, we know the exact number of bytes used by the columns and the validity
+   * bytes. We can use that value to branch over the size of the output to know which kernel to
+   * call.
+   *
+   * @param cudfUnsafeRow the unsafeRow object that represents the Spark SQL Schema.
+   * @return true if {@link CudfUnsafeRow#getFixedWidthSizeInBytes()} is less than
+   *         {@link #JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED} which is 1528 Bytes.
+   */
+  public static boolean fitsOptimizedConversion(
+      CudfUnsafeRow cudfUnsafeRow) {
+    return !cudfUnsafeRow.isVariableWidthSchema() &&
+        cudfUnsafeRow.getVariableSizeDataOffset() < JCUDF_MAX_FIXED_ROW_SIZE_FITS_OPTIMIZED;
```

Comment:
This seems odd to say "if it isn't variable width and the offset to variable width data is...". I get the idea here, but is there a method or should there be a method for getting the length of the fixed-width data + validity? This is it, obviously, but it seems odd to call it when it isn't a variable-length data schema. Not picky about this, just a talking point.

Reply:
changed to `getFixedWidthInBytes()`